portal-kernel/src/com/liferay/portal/kernel/util/LocaleUtil.java

```diff
@@ -279,19 +265,51 @@ else if (languageId.equals("zh-Hant-TW")) {
 			}
 		}
 
-		if (validate && !LanguageUtil.isAvailableLocale(locale)) {
-			locale = null;
+		_locales.put(languageId, locale);
+
+		return locale;
+	}
+
+	private Locale _fromLanguageId(String languageId, boolean validate) {
+		return _fromLanguageId(languageId, validate, true);
+	}
+
+	private Locale _fromLanguageId(
+		String languageId, boolean validate, boolean useDefault) {
 
-			if (_log.isWarnEnabled()) {
-				_log.warn(languageId + " is not a valid language id");
+		if (languageId == null) {
+			if (useDefault) {
+				return _getDefault();
 			}
-		}
-		else {
-			_locales.put(languageId, locale);
+
+			return null;
 		}
 
-		if ((locale == null) && useDefault) {
-			locale = _locale;
+		Locale locale = _fromLanguageId(languageId);
+
+		if (validate) {
+			boolean languageCode = false;
+
+			if ((languageId.indexOf(CharPool.UNDERLINE) < 0) &&
+				(languageId.indexOf(CharPool.MINUS) < 0)) {
+
+				languageCode = true;
+			}
+
+			if ((languageCode &&
+				 !LanguageUtil.isAvailableLanguageCode(languageId)) ||
+				(!languageCode && !LanguageUtil.isAvailableLocale(locale))) {
+
+				if (_log.isWarnEnabled()) {
+					_log.warn(languageId + " is not a valid language id");
+				}
+
+				if (useDefault) {
+					return _locale;
```

Comment:
This looks a bit weird, at https://github.com/liferay-core-infra/liferay-portal/pull/743/files#diff-19fc95663d567d9c82d2311555a10f4342a170408c701037791a1615328fbdcfR282 it returns "_getDefault()" which read the threadlocal, but here it only reads the field.
I know this pull did not change the logic for it, but since you are correcting the behavior for consistency, I would expect an adjustment for this too.